pkg/controller/instance/controller_status.go

```diff
@@ -16,34 +16,150 @@ package instance
 
 import (
 	"context"
+	"encoding/json"
 	"fmt"
-	"time"
 
-	corev1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
 
 	"github.com/kro-run/kro/api/v1alpha1"
+	"github.com/kro-run/kro/pkg/apis"
 	"github.com/kro-run/kro/pkg/requeue"
 )
 
-func createCondition(conditionType v1alpha1.ConditionType, status corev1.ConditionStatus, reason, message string, generation int64) map[string]interface{} {
-	return map[string]interface{}{
-		"type":               string(conditionType),
-		"status":             string(status),
-		"reason":             reason,
-		"message":            message,
-		"lastTransitionTime": time.Now().Format(time.RFC3339),
-		"observedGeneration": generation,
+const (
+	Ready           = "Ready"
+	InstanceManaged = "InstanceManaged"
+	GraphResolved   = "GraphResolved"
+	ResourcesReady  = "ResourcesReady"
+)
+
+var instanceConditionTypes = apis.NewReadyConditions(InstanceManaged, GraphResolved, ResourcesReady)
+
+// unstructuredConditionAdapter adapts an unstructured.Unstructured to implement apis.Object
+type unstructuredConditionAdapter struct {
+	*unstructured.Unstructured
+}
+
+// GetConditions implements apis.Object interface
+func (u *unstructuredConditionAdapter) GetConditions() []v1alpha1.Condition {
+	if conditions, found, err := unstructured.NestedSlice(u.Object, "status", "conditions"); err == nil && found {
+		// Marshal the conditions slice to JSON and then unmarshal to []v1alpha1.Condition
+		conditionsJSON, err := json.Marshal(conditions)
+		if err != nil {
+			return []v1alpha1.Condition{}
+		}
+
+		var result []v1alpha1.Condition
+		if err := json.Unmarshal(conditionsJSON, &result); err != nil {
+			return []v1alpha1.Condition{}
+		}
+		return result
 	}
+	return []v1alpha1.Condition{}
+}
+
+// SetConditions implements apis.Object interface
+func (u *unstructuredConditionAdapter) SetConditions(conditions []v1alpha1.Condition) {
+	// Marshal the conditions to JSON and then unmarshal to interface{} slice
+	conditionsJSON, err := json.Marshal(conditions)
+	if err != nil {
+		return // Fail silently - could log this in the future
+	}
+
+	var conditionsInterface []interface{}
+	if err := json.Unmarshal(conditionsJSON, &conditionsInterface); err != nil {
+		return // Fail silently - could log this in the future
+	}
+
+	if err := unstructured.SetNestedSlice(u.Object, conditionsInterface, "status", "conditions"); err != nil {
+		return // Fail silently - could log this in the future
+	}
```

Comment:
a part of things that this is worthy of a panic. basically this should never happen, we own the full logic of conditions